### DISCUSSION TITLE
Using SQL Views on Prisma View

### DISCUSSION THREAD

**Discussion by Niedman** [prisma/prisma#24082] [Q&A]
### Question

Hi Everyone,

I've followed the official documentation(https://www.prisma.io/docs/orm/prisma-schema/data-model/views) of Prisma in order to fetch data from a view, however I'm getting `The table public.BalanceTransactionView does not exist in the current database.`

If I use $queryraw i'm able to fetch the data and don't get that error but when i use the client like: `prisma.BalanceTransacionView.findMany()` it states that the The table public.BalanceTransactionView does not exist in the current database.

I've added the view to the prisma.schema and run prisma generate so I can use the prisma client to the new generated view and I can see that it recognizes it but when i deploy it it just gives the error above. From what i understood from the documentation we don't need to create a migrate so I'm a bit unsure why does it not get the view.

The view on the scheme is defined: 

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["views"]
}
```

```
view BalanceTransactionView {
  id                     String   @unique
...
}
```

Thanks in advance.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @Niedman 👋 

Thank you for raising this question.

Currently, to use `Views` in Prisma, you will need to add it as part of a migration using the `migrate dev --create-only` command and then manually add the SQL for views to your migration file. Alternatively, you can create views manually in the database. and then introspect it using `npx prisma db pull`. 

Defining a View in the schema file and running `prisma generate` does not create or run any SQL related to views which is why you are receiving the error message `does not exist in the current database`.

**Reply by Niedman** [reply]
Thanks for the reply @ludralph !

I've made the second option which was creating the view on the database and then did a npx prisma db pull but it deletes the view from the prisma scheme. Is that supposed to happen? 

I will give it a try with a migration and see how it goes :) I will post when it's fixed!

**Reply by nikolasburk** [reply]
Did it work in the end for you, @Niedman?

**Reply by aquiles419** [reply]
@Niedman , Were you able to query a view?

**Reply by Niedman** [reply]
@nikolasburk @aquiles419 yes I was. It’s pretty much the same as a table and it works like a charm 😊 the view will not be created with the migration so you need to manually create it but apart from this is pretty much straightforward.

**Reply by aquiles419** [reply]
@Niedman , Can you describe the steps to me? I tried to make a query here, I made a view in a SQL Server, I pulled it into my application, I can even get the view in my prism, it generates as if it were a new entity, but it does not open the methods like findMany ..... I am using version 5.18.0

**Reply by Niedman** [reply]
@aquiles419 sure :) 

Well i've added the view to the schema.prisma using view instead of model for tables. 
Also you need to add `previewFeatures = ["views"]` in the schema generator.
Check this page: https://www.prisma.io/docs/orm/prisma-schema/data-model/views

Reloaded the schema so that my code is aware of the view.(I ran this on a local db(dockerized) to be sure but it's optional) 

And then I literally used the same way as a model.

schema.prisma
```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["views"]
}
```
````
view balancetransactionview {
  id                     String    @unique
etc
etc
}
```

usage:
```
try {
      const [prismaBalanceTransactionsView, total] =
        await this.prisma.$transaction([
          this.prisma.balancetransactionview.findMany({
            take: parsedQuery.take,
            skip: parsedQuery.skip,
            cursor: parsedQuery.cursor,
            where: parsedQuery.where,
            orderBy: parsedQuery.sort,
          }),
          this.prisma.balancetransactionview.count({
            where: parsedQuery.where,
          }),
        ]);
}
```

hope it helps :)

**Reply by aquiles419** [reply]
@Niedman , Your explanation is very cool, thank you. But I still have the same error, `TypeError: Cannot read properties of undefined (reading 'findMany')`, I've tried everything, I've read the complete documentation, I've followed all the steps, my view exists in schema.prisma, but I still can't search!

**Reply by Niedman** [reply]
@aquiles419 have you generated the schema again? When you generate the schema again a new folder should appear on your prisma folder. Something like this:

<img width="278" alt="Screenshot 2024-09-02 at 09 11 34" src="https://github.com/user-attachments/assets/323b8feb-5c14-4064-8eab-c8b5b9f2cbd4">

I remember that I struggled a bit with the same. My advise is to change your datasource url to a local docker instance and generate the schema again an that should create a the view folder.

```
datasource db {
  provider = "postgresql"
  url      = "<your docker connection string>"
}
```

**Reply by aquiles419** [reply]
@Niedman You were right, I had forgotten to remove the comment: 
`/// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by Prisma Client. `

and the  `@@ignore `

I put a `@unique` in the id of the unique identifier in the table (view), and it worked! Thank you very much.

**Reply by amitbeck** [reply]
> @aquiles419 have you generated the schema again? When you generate the schema again a new folder should appear on your prisma folder. Something like this:
> 
> <img alt="Screenshot 2024-09-02 at 09 11 34" width="278" src="https://private-user-images.githubusercontent.com/8912378/363592108-323b8feb-5c14-4064-8eab-c8b5b9f2cbd4.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI2ODEzOTQsIm5iZiI6MTc2MjY4MTA5NCwicGF0aCI6Ii84OTEyMzc4LzM2MzU5MjEwOC0zMjNiOGZlYi01YzE0LTQwNjQtOGVhYi1jOGI1YjlmMmNiZDQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMDlUMDkzODE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGM1N2Y1OWZmY2JjNTliZDcyZTJhMTBjMzljNTllMmQ5ZWYxOWEwYWI4MjFlYTRlODA4ZDA4MjExOTg3NzhkNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.GeoaFendRZK7ChOSMd9qjPlx7O4PJjvyy0hyEmQdAeg">
> I remember that I struggled a bit with the same. My advise is to change your datasource url to a local docker instance and generate the schema again an that should create a the view folder.
> 
> ```
> datasource db {
>   provider = "postgresql"
>   url      = "<your docker connection string>"
> }
> ```

@Niedman Any idea what the purpose of this folder is, or why it is even needed?
